addons/website_sale_coupon/controllers/main.py

```diff
@@ -28,6 +28,56 @@ def cart(self, **post):
         order.recompute_coupon_lines()
         return super(WebsiteSale, self).cart(**post)
 
+    @http.route(['/coupon/<string:code>'], type='http', auth='public', website=True, sitemap=False)
+    def activate_coupon(self, code, **post):
+        error_status = self._try_coupon(code)
+
+        if error_status:
+            #TODO Redirection to an invalid coupon page might be more user-friendly?
+            pass
+
+        return request.redirect('/shop')
+
+    @http.route()
+    def cart_update(self, **kw):
+        result = super(WebsiteSale, self).cart_update(**kw)
+
+        self._try_coupon_from_session()
+        return result
+
+    @http.route()
+    def cart_update_json(self, **kw):
+        result = super(WebsiteSale, self).cart_update_json(**kw)
+
+        self._try_coupon_from_session()
+        return result
+
+    def _try_coupon_from_session(self):
+        coupon_data = request.session.get('website_sale_coupon_activated_coupon', None)
+        error_status = {}
+
+        if coupon_data:
+            error_status = self._try_coupon(coupon_data['coupon_code'])
+
+            if not error_status:
+                request.session['website_sale_coupon_activated_coupon'] = None
+
+        return error_status
+
+    def _try_coupon(self, code):
+        order = request.website.sale_get_order()
```

Comment:
All this can go directly on order model and be called in `recompute_coupon_lines` and route overide removed, as it will be called by every endpoints and we will be sure we don't miss one (add to cart, cart update, payment steps etc)

Reply:
I don't think I have access to the session from `recompute_coupon_lines`. Should I store the coupon on a field on the `sale.order` instead?